File: src/onjava/codility/arrays/CyclicRotations.java
package onjava.codility.arrays;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

/**
 * Site: Codility
 * Problem: Cyclic Rotations
 *
 * User: Luis Villaca
 * Date: 20/04/2020
 */
public class CyclicRotations {
    public int[] solution(int[] A, int N) {
        //CyclicRotation

        int rotations = N%A.length;

        int[] intNewArray = new int[A.length];

        // calculate the new index for this element
        for (int i=0; i<A.length;i++) {
            int newIndex = (i + rotations) % A.length;
            intNewArray[newIndex]=A[i];
        }
        return intNewArray;
    }


    @DataProvider(name = "tests")
    public Object [][] testData() {
        return new Object [][] {
                new Object [] { new int [] { 3, 8, 9,  7, 6 },  3,  new int [] {  9, 7, 6,  3, 8 } },
                new Object [] { new int [] {       0,  0, 0 },  1,  new int [] {        0,  0, 0 } },
                new Object [] { new int [] {    1, 2,  3, 4 },  4,  new int [] {     1, 2,  3, 4 } },
                new Object [] { new int [] {    1, 2,  3, 4 },  5,  new int [] {     4, 1,  2, 3 } },
                new Object [] { new int [] {   -1, 2, -3, 4 }, 10,  new int [] {    -3, 4, -1, 2 } },
                new Object [] { new int [] {   -1, 2, -3, 4 }, 99,  new int [] {    2, -3, 4, -1 } },
                new Object [] { new int [] {   -1, 2, -3, 4 }, 100,  new int [] {   -1, 2, -3, 4 } },
        };
    }

    @Test(dataProvider = "tests")
    public void verify(int [] pA, int pK, int [] pExpectedRotatedA) {
        int [] actualRotatedA = solution(pA, pK);
        Assert.assertEquals(actualRotatedA.length, pExpectedRotatedA.length);
        for(int i=0; i<pExpectedRotatedA.length; i++) {
            Assert.assertEquals(actualRotatedA[i], pExpectedRotatedA[i]);
        }
    }
}

File: src/onscala/codility/arrays/CyclicRotations.scala
package onscala.codility.arrays

import org.scalatest.Assertions
import org.testng.Assert._
import org.testng.annotations.{DataProvider, Test}

/**
  * Site: Codility
  * Problem: Cyclic Rotations
  *
  * User: Luis Villaca
  * Date: 24/05/2020
  */
class CyclicRotations extends Assertions {

  def solution(a: Array[Int], k: Int): Array[Int] = {
      if (a.isEmpty) a
      else {
        val rotations = k % a.length
        val intNewArray = Array.fill(a.length)(0)

        // calculate the new index for this element
        intNewArray.indices.foreach(ind => {
          val newIndex = (ind + rotations) % a.length
          intNewArray(newIndex) = a(ind)
        })
        intNewArray
      }
    }

  //Evaluating
  @DataProvider(name= "tests")
  def testData() = {
    Array(
      Array[Any](Array(3, 8, 9,  7, 6), 3, Array(9, 7, 6,  3, 8)),
      Array[Any](Array(0,  0, 0), 1, Array(0,  0, 0)),
      Array[Any](Array(1, 2,  3, 4), 4, Array(1, 2,  3, 4)),
      Array[Any](Array(1, 2,  3, 4), 5, Array(4, 1,  2, 3)),
      Array[Any](Array(-1, 2, -3, 4), 10, Array(-3, 4, -1, 2)),
      Array[Any](Array(-1, 2, -3, 4), 99, Array(2, -3, 4, -1)),
      Array[Any](Array(-1, 2, -3, 4), 100, Array(-1, 2, -3, 4))
    )
  }

  @Test(dataProvider = "tests")
  def verify(arrayInput: Array[Int], intInput: Int, expectedOutput: Array[Int]): Unit = {
    val actualOutput = solution(arrayInput, intInput)
    assertEquals(actualOutput.length, arrayInput.length)
    actualOutput.indices.foreach(ind =>
      assertEquals(actualOutput(ind), expectedOutput(ind))
    )
  }

}







